k8s/pkg/apis/longhorn/v1beta2/objectendpoint.go

```diff
@@ -0,0 +1,178 @@
+	UnmapMarkSnapChainRemoved UnmapMarkSnapChainRemoved `json:"unmapMarkSnapChainRemoved,omitempty"`
+	// Replica soft anti affinity of the volume. Set enabled to allow replicas to be scheduled on the same node.
+	// +optional
+	ReplicaSoftAntiAffinity ReplicaSoftAntiAffinity `json:"replicaSoftAntiAffinity,omitempty"`
+	// Replica zone soft anti affinity of the volume. Set enabled to allow replicas to be scheduled in the same zone.
+	// +optional
+	ReplicaZoneSoftAntiAffinity ReplicaZoneSoftAntiAffinity `json:"replicaZoneSoftAntiAffinity,omitempty"`
+	// Replica disk soft anti affinity of the volume. Set enabled to allow replicas to be scheduled in the same disk.
+	// +optional
+	ReplicaDiskSoftAntiAffinity ReplicaDiskSoftAntiAffinity `json:"replicaDiskSoftAntiAffinity,omitempty"`
+	// +optional
+	FromBackup string `json:"fromBackup,omitempty"`
+	// +optional
+	RecurringJobSelector []VolumeRecurringJob `json:"recurringJobSelector,omitempty"`
+	// +optional
+	DiskSelector []string `json:"diskSelector,omitempty"`
+	// +optional
+	NodeSelector []string `json:"nodeSelector,omitempty"`
+	// +kubebuilder:validation:Enum=v1;v2
+	// +optional
+	BackendStoreDriver BackendStoreDriverType `json:"backendStoreDriver,omitempty"`
+}
+
+type ObjectEndpointSpec struct {
+	// Credentials contain a pair of access and secret keys that are used to seed
+	// the object endpoint.
+	//
+	// +optional
+	Credentials ObjectEndpointCredentials `json:"credentials"`
+
+	// Maybe we don't need user to specify StorageClass because we are doing static provisioning now
+	// The name of a storage class. The backing storage is going to be a new
+	// volume from that storage class.
+	//
+	// +optional
+	//StorageClass string `json:"storageClassName"`
+
+	// S3GW Image
+	// +optional
+	Image string `json:"image"`
+
+	// S3GW UI Image
+	// +optional
+	UiImage string `json:"UiImage"`
+
+	// +optional
+	FromVolumeBackup string `json:"fromVolumeBackup"`
+
+	// VolumeParameters holds the parameters for the backing Longhorn volume of this ObjectEndpoint
+	// +optional
+	VolumeParameters ObjectEndpointVolumeParameters `json:"volumeParameters"`
+
+	// The initial size of the volume to provision. This size together with the
+	// storage class will be used to automatically create a suitable volume where
+	// the objects will be stored.
+	//
+	// +optional
+	Size resource.Quantity `json:"size"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:shortName={lhoe}
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Name",type=string,JSONPath=`.metadata.name`
+// +kubebuilder:printcolumn:name="State",type=string,JSONPath=`.status.state`,description="The state of object endpoint"
+// +kubebuilder:printcolumn:name="Endpoint",type=string,JSONPath=`.status.endpoint`,description="The endpoint address"
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
```

Comment:
@m-ildefons 

> This may create naming conflicts with non-namespaced resources like PersistentVolume that need to be created automatically by the controller.

It is by design that the Longhorn system should only be deployed once in the cluster. Because each objectEnpoint name is unique in the namespace of Longhorn system, there shouldn't be naming conflicts for PV

> It also breaks the ownership references if the controller creates the resources in the longhorn-system namespace but the ObjectEndpoint resource resides in another namespace.

The controller and ObjectEnpoint CRs are always in the same namespace. I think it is not possible for them to be in different namespaces with the current implementation. WDYT?

All Longhorn CRD has namespace scope. I think the ObjectPoint should have namespace scope similar to all other Longhorn resources unless there is a convincing reason for not doing so. When the users delete longhorn-system namespace, it will automatically delete all resources inside it (including ObjectEndpoints)

Reply:
> The controller and ObjectEnpoint CRs are always in the same namespace
If the ObjectStore (ex ObjectEndpoint) resource is namespaced, a user may create an ObjectStore resource in a different namespace. Therefore we can not assume that the ObjectStore is always in the longhorn-system namespace, if it is a namespaced resource.

We can not have ownership references of namespaced objects (like Deployments) to namespaced objects in different namespaces. That is not allowed by the K8s API. Therefore if the ObjectStore must be a namespaced object we must enforce that it can only be created in the longhorn-system namespace.

Initially, I tried to have the resource namespaced, but I had to make it cluster scoped because of the ownership references that are used to manage the dependent resources.
However, if I can enforce that the resource can only be created in the longhorn-system namespace, the resource can be namespaced.

Reply:
Yes, we can and currently enforce that all namespaced resources (including deployment, services, secret, PVC) of the Longhorn system are inside a single namespace (usually that namespace is named as longhorn-system depend on how user deploy Longhorn). So, I think it ok to make ObjectStore CR to have namespaced scope